### PR TITLE
feat: enable multithreading for ZIO Native support

### LINKED ISSUE #9056: Enable multithreading for ZIO Native support
This will require ditching the versions of many structures copy/pasted from Scala.js and using true multi-threaded implementations, potentially from the JDK if they have Scala Native analogues.

### PR DESCRIPTION
**Pull request by varshith257**
This PR addresses issue #9056 by enabling multithreading support in Scala Native. The goal was to replace Scala.js copy/pasted structures with true multithreaded implementations enabling Scala Native to fully support concurrent operations across all tests.

### Why We Chose ScheduledExecutorService and Optimal Thread Pool Size

**Rationale for Choosing ScheduledExecutorService:**

`ScheduledExecutorService` is a standard Java utility that provides us with a way to schedule tasks for execution in a thread pool. It supports both fixed-delay and fixed-rate executions which are crucial for accurately timing operations in a concurrent environment. Being a part of the Java standard library, `ScheduledExecutorService` is a well-tested and widely-used solution.

Reference: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledThreadPoolExecutor.html

### Key Changes:

The core enhancement involves the implementation of a multithreaded scheduler using `ScheduledThreadPoolExecutor`. This executor is configured with a core pool size of 2 threads providing a balance between resource utilization and performance. The pool size can be adjusted based on workload requirements.

`timeoutWithTrace`  method is added to schedule tasks after a specified delay. This method uses the `ScheduledThreadPoolExecutor` to execute a callback after a given duration.

`repeatWithTrace`  method enables the periodic execution of tasks with fixed intervals. This is needed for tasks that need to run at regular intervals without overlap or concurrency issues.

Also made `ConcurrentMetricHooksPlatformSpecific` implementation of Native by making it fully thread-safe. The changes involve updating how metrics (counters, gauges, histograms, summaries and frequencies) are managed in a concurrent environment to ensure that all operations are atomic and consistent.

fixes #9056 
/claim #9056
/split @asr2003

### CODE REVIEW

**Comment by varshith257** [comment]
Interesting! All native tests are passed locally and reporting here fails. Will fix it

**Comment by varshith257** [comment]
@jdegoes Here we go!

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala [hunk 1]
```diff
@@ -52,102 +53,53 @@ private[zio] trait ClockPlatformSpecific {
...
-        !ptrOfPtr = fromRawPtr[Byte](rawptr)
-      } else {
-        !ptrOfPtr = null
-      }
-    }
-    private val onCloseCB: CloseCB = (handle: UVHandle) => {
-      stdlib.free(handle)
-    }
+  // Multi-threaded scheduler using ScheduledExecutorService
+  val scheduler: ScheduledExecutorService = Executors.newScheduledThreadPool(4)
```
Comment (sideeffffect) [inline core/native/src/main/scala/zio/ClockPlatformSpecific.scala]: Shouldn't this reflect the number of available cores?
```
java.lang.Runtime.getRuntime().availableProcessors()
```

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala [hunk 2]
```diff
@@ -52,102 +53,53 @@ private[zio] trait ClockPlatformSpecific {
...
+        TimeUnit.MILLISECONDS
+      )
+      new Timer(scheduledFuture)
     }
 
-    def delay(duration: FiniteDuration): Future[Unit] = {
-      val promise = scala.concurrent.Promise[Unit]()
-      timeout(duration)(() => promise.success(()))
-      promise.future
+    // Original method retained for backward compatibility
```
Comment (sideeffffect) [inline core/native/src/main/scala/zio/ClockPlatformSpecific.scala]: Do we even want/need to keep backwards compatibility?

**Review by kyri-petrou — CHANGES_REQUESTED** [review]
I'm finding this a little bit too difficult to believe that it's working with these number of changes. There are _a lot_ more thread-unsafe implementations in ZIO used for Scala Native. Also, this PR still uses the `native-loop-core` library, which even [according to the library author](https://discord.com/channels/632150470000902164/635668881951686686/1272966616975409253) isn't meant to work with multithreading.

Frankly, I have no idea why CI is passing (when it really shouldn't), but I have a strong feeling is because multi-threading isn't actually being used due to `native-loop-core`. I think as part of this PR it needs to be removed as a dependency and rely on Java Threads / Executors for multi-threading instead.

File: project/BuildHelper.scala [hunk 3]
```diff
@@ -242,7 +242,7 @@ object BuildHelper {
   )
 
   def nativeSettings = Seq(
-    nativeConfig ~= { _.withMultithreading(false) },
+    nativeConfig ~= { _.withMultithreading(true) },
```
Comment (sideeffffect) [inline project/BuildHelper.scala]: Maybe we can even remove this line. It defaults to true anyway.

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala [hunk 4]
```diff
@@ -52,102 +53,53 @@ private[zio] trait ClockPlatformSpecific {
...
- * Copy-pasted from scala.scalanative.loop
- *
- * This is temporary until we figure out why using it directly raises errors
- */
 private object ClockPlatformSpecific {
-  import scala.collection.mutable
-  import scala.concurrent.Future
-  import scala.scalanative.annotation.alwaysinline
-  import scala.scalanative.libc.stdlib
-  import scala.scalanative.loop.EventLoop
```
Comment (sideeffffect) [inline core/native/src/main/scala/zio/ClockPlatformSpecific.scala:65]: Are still using even anything from `scala-native-loop`at this point? 
If not, we could remove the dependency.

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala [hunk 5]
```diff
@@ -52,102 +53,53 @@ private[zio] trait ClockPlatformSpecific {
...
+      } yield ()
+
+    // New method added with Trace parameter for new functionality
+    private def timeoutWithTrace(duration: FiniteDuration)(callback: () => Unit)(implicit trace: Trace): Timer = {
+      val scheduledFuture = scheduler.schedule(
+        new Runnable {
+          override def run(): Unit = callback()
+        },
+        duration.toMillis,
+        TimeUnit.MILLISECONDS
```
Comment (sideeffffect) [inline core/native/src/main/scala/zio/ClockPlatformSpecific.scala]: Would it make sense to have higher precision, like Microseconds?

**Review by sideeffffect — COMMENTED** [review]
Thank you very much for this work 🙏

**Comment by kyri-petrou** [comment]
To compliment my previous message: `ZSTMLockSupport` still [uses the same implementation](https://github.com/zio/zio/blob/series/2.x/core/js-native/src/main/scala/zio/stm/ZSTMLockSupport.scala) as ScalaJS. There is no way ZSTM tests would be passing without locks if we had true parallelism

**Comment by varshith257** [comment]
> Frankly, I have no idea why CI is passing (when it really shouldn't), but I have a strong feeling is because multi-threading isn't actually being used due to `native-loop-core`. I think as part of this PR it needs to be removed as a dependency and rely on Java Threads / Executors for multi-threading instead.

 I agree that using the `native-loop-core library` while trying to enable multithreading could be contradictory. The initial focus was on gradually introducing multithreading in areas that needed it the most. Removing dependencies like native-loop-core will be the next step to ensure a clean and truly multithreaded environment to rely on Java’s thread and executor mechanisms.

**Comment by kyri-petrou** [comment]
@varshith257 I disagree. If the author of `native-loop-core` says "It doesn't work with multi-threading" then we really shouldn't be using this library with multi-threading enabled.

Have you tried removing the dependency? It's only used in `OneShot`. What happens when you remove it?

**Comment by varshith257** [comment]
Yes! It's working fine with removing it. I just doesn't pushed changes tested locally just now. 

For replacement of current `ZSTMLockSupport` of ScalaJS for Native too, I have gone through how JVM handled through `ReentrantLock` but Native doesn't support it. I think we need to adapt it using atomic operations and locks that are compatible with Scala Native env.

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala (hunk 4 above)
Comment (varshith257) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala:65]: Just removed it!

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala (hunk 5 above)
Comment (varshith257) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala]: Yes it makes sence of increasing the precision of time units to microseconds can make more precise in scheduling tasks

**Comment by kyri-petrou** [comment]
> I have gone through how JVM handled through ReentrantLock but Native doesn't support it.

It does since v0.5.0: https://github.com/scala-native/scala-native/blob/d5d65e755a4fb40a1c562d50b046384000232b23/javalib/src/main/scala/java/util/concurrent/locks/ReentrantLock.scala#L148

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala (hunk 2 above)
Comment (varshith257) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala]: I need to check on this. I have faced some mima errors locally so preserved it. Well, I will look into it.

**Comment by kyri-petrou** [comment]
There is something very wrong going on here. There are failures in the tests in CI and these are not being reported as an exit code 1 to the SBT runner: https://github.com/zio/zio/actions/runs/10417903062/job/28853002331#step:7:1542

image

**Comment by varshith257** [comment]
This is another test failing along with the above attached after enabling multithreading for Scala Native.

image


No idea, what caused them to fail and doesn't report to SBT runner. Any guidance on this?

cc: @ghostdogpr @jdegoes [Tagged to get any guidance on this behaviour, maybe your experience could be valuable here]

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala (hunk 2 above)
Comment (varshith257) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala]: They needed for happy path of mima

**Comment by varshith257** [comment]
> To compliment my previous message: `ZSTMLockSupport` still [uses the same implementation](https://github.com/zio/zio/blob/series/2.x/core/js-native/src/main/scala/zio/stm/ZSTMLockSupport.scala) as ScalaJS. There is no way ZSTM tests would be passing without locks if we had true parallelism

@kyri-petrou Yes, the tests failed indeed for using ScalaJS ZSTM Lock. After using ZSTMLockSupport using ReetrantLock similar to 
JVM, the tests are consistently passing without any error. Check here

image
https://github.com/zio/zio/actions/runs/10420377547/job/28860271411#step:7:1436

image

https://github.com/zio/zio/actions/runs/10420377547/job/28860271411#step:7:2217


Thanks a lot for this!

File: core/native/src/main/scala-2/zio/stm/ZSTMLockSupport.scala
Comment (sideeffffect) [inline core/native/src/main/scala-2/zio/stm/ZSTMLockSupport.scala:1]: This is identical to the `jvm` variant. Just put move the file from `jvm` to `jvm-native` (as before there was in `js-native`).

**Review by sideeffffect — COMMENTED** [review]
I think that before we merge this, we should think carefully about the separation between the `jvm` code and `native` code.
Scala Native has already bridged the gap quite a lot, many of the common standard library classes are already implemented.

Maybe we'll realize that most of the code could live in `jvm-native` and `js` -- i.e. JVM and Native sharing most (if not all) implementation.

What do you think?

**Comment by kyri-petrou** [comment]
@varshith257 the test that was failing was not a ZSTM test. It's just a coincidence that it's now passing.

File: core/native/src/main/scala-2/zio/stm/ZSTMLockSupport.scala
Comment (varshith257) [reply core/native/src/main/scala-2/zio/stm/ZSTMLockSupport.scala:1]: Done

File: core/native/src/main/scala/zio/internal/OneShot.scala
Comment (sideeffffect) [inline core/native/src/main/scala/zio/internal/OneShot.scala:1]: I already made the general point, but here specifically -- do we really need separate implementation of OneShot for `jvm` and `native`?

Comment (varshith257) [reply core/native/src/main/scala/zio/internal/OneShot.scala:1]: I think we don't need separate ones because native already supports `Reentrant` and `Condition` Locks which are used by Oneshot of jvm. So we can move it to `jvm-native`.

**Comment by varshith257** [comment]
@kyri-petrou @sideeffffect Do you have any idea about why CI Scala 2.12.x is failing? It started failing after restructuring to `jvm-native`.

File: core/native/src/main/scala/zio/internal/OneShot.scala
Comment (varshith257) [reply core/native/src/main/scala/zio/internal/OneShot.scala:1]: Done!

**Comment by kyri-petrou** [comment]
@varshith257 there is still a lot of work to be done in this PR, I would argue it's better put it to draft until the work is done.

Along with what @sideeffffect suggested, first, all the code under `/jvm` that is compatible with Scala Native should be moved under `/jvm-native`. Then we need a proper Scheduler implementation (similar to ZScheduler). Maybe the jvm one is fully compatible I don't know. And finally test failures need to be properly reported.

By the end of it, almost all code currently under `/native` in the `series/2.x` branch should be deleted or changed

**Comment by varshith257** [comment]
Okay! Will follow up with @sideeffffect

File: concurrent/shared/src/test/scala/zio/concurrent/ConcurrentMapSpec.scala [hunk 6]
```diff
@@ -256,7 +256,7 @@ object ConcurrentMapSpec extends ZIOSpecDefault {
             bRes <- map.get(2)
             cRes <- map.get(3)
           } yield assertTrue(
-            aRes.get == "A",
+            aRes.get == "B",
```
Comment (bishabosha) [inline concurrent/shared/src/test/scala/zio/concurrent/ConcurrentMapSpec.scala]: why?

Comment (bishabosha) [reply concurrent/shared/src/test/scala/zio/concurrent/ConcurrentMapSpec.scala]: ok I assume its because of some deliberate attempt to make it fail given the commit name

Comment (varshith257) [reply concurrent/shared/src/test/scala/zio/concurrent/ConcurrentMapSpec.scala]: Yes, it has been made to make it fail intending to report failure of CI

Comment (varshith257) [reply concurrent/shared/src/test/scala/zio/concurrent/ConcurrentMapSpec.scala]: @bishabosha Have you observed the behavior of Native CI. The rest are reporting it's failure and Native passed 🤒

Comment (varshith257) [reply concurrent/shared/src/test/scala/zio/concurrent/ConcurrentMapSpec.scala]: It deeply rooted without multithreading too

**Comment by varshith257** [comment]
@kyri-petrou @sideeffffect It's now set to go 🚀

**Comment by varshith257** [comment]
> And finally test failures need to be properly reported.

It is not part of PR as discussed in Discord but exist before too. I or maybe @asr2003 will create new issue for it 

Ref : - https://github.com/zio/zio/actions/runs/10430627186/job/28889380600?pr=9124#step:7:1824


- https://github.com/zio/zio/pull/9124

File: core/js/src/main/scala-3/zio/stm/ZSTMLockSupport.scala
Comment (sideeffffect) [inline core/js/src/main/scala-3/zio/stm/ZSTMLockSupport.scala:1]: Why does there now need to be a Scala 3 and Scala 2 specific `ZSTMLockSupport` for Scala.js?

Comment (varshith257) [reply core/js/src/main/scala-3/zio/stm/ZSTMLockSupport.scala:1]: It's difference between inline keyword that supported in Scala3 but not in Scala2

File: core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala [hunk 7]
```diff
@@ -317,7 +317,10 @@ private final class ZScheduler(autoBlocking: Boolean) extends Executor {
                       if (nRunnables > 0) {
                         val iter = runnables.iterator
                         runnable = iter.next()
-                        if (nRunnables > 1) localQueue.offerAll(iter, nRunnables - 1)
+                        if (nRunnables > 1) {
+                          val iterable = iter.toIterable
```
Comment (sideeffffect) [inline core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala:321]: Where is `iterable` used? If nowhere, why not `val _ = ...`?

Comment (varshith257) [reply core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala:321]: Yes, can make it though 😅

File: core/native/src/main/scala/zio/RuntimePlatformSpecific.scala [hunk 8]
```diff
@@ -38,7 +39,10 @@ private[zio] trait RuntimePlatformSpecific {
   final val defaultReportFatal: Throwable => Nothing =
     (t: Throwable) => {
       t.printStackTrace()
-      throw t
+      try {
+        java.lang.System.exit(-1)
```
Comment (sideeffffect) [inline core/native/src/main/scala/zio/RuntimePlatformSpecific.scala]: I suspect this isn't correct. This would bring the whole process down. But that's not what it should do.
The ZIO runtime could be only a part of what runs in that given process.

Comment (varshith257) [reply core/native/src/main/scala/zio/RuntimePlatformSpecific.scala]: I just realised that it is done to test its behaviour anyhow will remove it

File: test/native/src/main/scala/zio/test/TestClockPlatformSpecific.scala [hunk 9]
```diff
@@ -16,19 +16,197 @@
...
+
+        def asScheduledExecutorService: ScheduledExecutorService = {
+          def compute[A](a: => A): Either[Throwable, A] =
+            try {
+              Right(a)
+            } catch {
+              case t: CancellationException => Left(t)
+              case t: InterruptedException  => Left(t)
+              case t: OutOfMemoryError      => throw t
+              case t: Throwable             => Left(new ExecutionException(t))
```
Comment (sideeffffect) [inline test/native/src/main/scala/zio/test/TestClockPlatformSpecific.scala]:
```suggestion
              case NonFatal(t)              => Left(new ExecutionException(t))
```

**Comment by kyri-petrou** [comment]
@varshith257 Scala Native tests are still failing: https://github.com/zio/zio/pull/9124/checks#step:7:1243

**Comment by varshith257** [comment]
@kyri-petrou The test fail is of another PR to check the current Native CI behaviour of test fail reporting

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala [hunk 10]
```diff
@@ -52,102 +54,51 @@ private[zio] trait ClockPlatformSpecific {
...
-        !ptrOfPtr = fromRawPtr[Byte](rawptr)
-      } else {
-        !ptrOfPtr = null
-      }
-    }
-    private val onCloseCB: CloseCB = (handle: UVHandle) => {
-      stdlib.free(handle)
-    }
+  // Multi-threaded scheduler using ScheduledExecutorService
+  val scheduler: ScheduledExecutorService = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors())
```
Comment (sideeffffect) [inline core/native/src/main/scala/zio/ClockPlatformSpecific.scala]:
```suggestion
  val scheduler: ScheduledExecutorService = Executors.newScheduledThreadPool(java.lang.Runtime.getRuntime().availableProcessors())
```

**Comment by varshith257** [comment]
> @varshith257 Scala Native tests are still failing: https://github.com/zio/zio/pull/9124/checks#step:7:1243

@kyri-petrou I found that I have moved the test clock of JVM to native but indeed not needed.

It's passing, check here:

https://github.com/zio/zio/actions/runs/10432370240/job/28893006297?pr=9117#step:7:1458

File: core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala [hunk 11]
```diff
@@ -317,7 +317,10 @@ private final class ZScheduler(autoBlocking: Boolean) extends Executor {
...
+                          localQueue.offerAll(iterable)
+                        }
```
Comment (kyri-petrou) [inline core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala:323]: Why is this change needed?

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala
Comment (kyri-petrou) [inline core/native/src/main/scala/zio/ClockPlatformSpecific.scala:1]: Can't we use the `/jvm` implementation here?

**Review by kyri-petrou — COMMENTED** [review]
@varshith257 The changes you did have nothing to do with the failing (flaky) test. If you run the test many times, it's definitely going to fail again at some point. The problem is that some implementation somewhere is not thread-safe. The reason that it doesn't fail every time is because in some cases, the concurrent modification that cause the failures won't be executed at the exact same time. Some other times, it will.

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala
Comment (varshith257) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala:1]: Yes, we need a change inorder to support for jvm-native a minor change

File: core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala (hunk 11 above)
Comment (varshith257) [reply core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala:323]: Yes, we need a change inorder to support for jvm-native a minor change

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala
Comment (varshith257) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala:1]: Yes, we can but in order to test that flaky test you intended, I thought that its difference of timing and concurrency I have kept separate

Comment (kyri-petrou) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala:1]: @varshith257 `Clock` has nothing to do with that test. If it's affecting the outcome of the test, we need to look into the root cause, because it _really_  should not be affecting its outcome

Comment (varshith257) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala:1]: Yes, either ways it reporting the same😅

File: core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala (hunk 11 above)
Comment (kyri-petrou) [reply core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala:323]: I don't understand why, can you please elaborate? In either way, this will make the JVM runtime slower, which we cannot tolerate. The performance of the JVM runtime should not be affected in any way for another runtime

File: core/native/src/main/scala/zio/ClockPlatformSpecific.scala
Comment (varshith257) [reply core/native/src/main/scala/zio/ClockPlatformSpecific.scala:1]: @kyri-petrou What do you think it causing test fails? I suspect that Native multithreading isn't as performant as JVM. It can go underhood of strict memory resource constraints too and  same of concurrency model of Native is more restricted tha in JVM especially when running tests in parallel that could lead to memory corruption 

I think we have moved from every Single threaded structures of Native to mostly to support Multithreading env (thought I can't say perfectly). Maybe your experience on Zio could be valuable of what leading this to fail under load of Native tests fail

**Review by sideeffffect — COMMENTED** [review]
I think we're getting very close, thank you so very much again for your effort and perseverance in this endeavor :bow: 

I have these observation:

# 1)
We shouldn't need the change in `ZScheduler` -- why can't we have `localQueue.offerAll(iter, nRunnables - 1)` even for `native`?

# 2)
There are also these `native` specific implementations that are kept (instead of merging them with `jvm`). It would be nice to have an explanation for each of them, why then can't be merged with `jvm`:
 * ConcurrentMetricHooksPlatformSpecific
 * ClockPlatformSpecific
 * FiberPlatformSpecific
 * TaskPlatformSpecific
 * ZIOPlatformSpecific

**Comment by varshith257** [comment]
> # 1)
> We shouldn't need the change in `ZScheduler` -- why can't we have `localQueue.offerAll(iter, nRunnables - 1)` even for `native`?

When we used without change reporting this error by Native CI:
```
[error] /home/runner/work/zio/zio/core/jvm-native/src/main/scala/zio/internal/ZScheduler.scala:320:82: too many arguments (found 2, expected 1) for method offerAll: (as: Iterable[A1]): zio.Chunk[A1]
[error]                         if (nRunnables > 1) localQueue.offerAll(iter, nRunnables - 1)
[error]                    
```
It is due to `offerAll` method having a different signature in Native. With accepting of single arg as I made change it working compatible with both native and JVM 

 # 2)
> There are also these `native` specific implementations that are kept (instead of merging them with `jvm`). It would be nice to have an explanation for each of them, why then can't be merged with `jvm`:

> * ConcurrentMetricHooksPlatformSpecific

It lacks support for DoubleAdder and the custom AtomicDouble we use for the JVM.

> * ClockPlatformSpecific

Yes, indeed we can make it => Ported

> * FiberPlatformSpecific
> * TaskPlatformSpecific
> * ZIOPlatformSpecific

It lacks support for `CompletableFuture` and  `CompletionStage` concurrent structures in Native. Though I have ported some code of JVM which is compatible and supported

**Comment by sideeffffect** [comment]
Any idea why the test `modifying different keys doesn't retry the transaction` has failed? Is it flaky?
https://github.com/zio/zio/actions/runs/10435458359/job/28899444305?pr=9117#step:7:2793

**Comment by varshith257** [comment]
Yes, it is flaky. After detailed isolation of tests and discussions with the `scala-native` community, the culprit here is `libuv` which we are using in our ci for the event loop which isn't thread-safe

**Comment by sideeffffect** [comment]
> `libuv`

Are you sure. AFAIK only scala-native-loop was using libuv underneath. So now that we don't use scala-native-loop, we could also remove installing libuv from the CI jobs.

**Comment by kyri-petrou** [comment]
> AFAIK only scala-native-loop was using libuv underneath

Agreed

**Comment by varshith257** [comment]
@sideeffffect Yes as suggested by @lolgab's comment:

**Lorenzo Gabriele (lolgab) — Yesterday at 8:02 PM**
_Basically libuv's API is not thread safe. So you need to make sure you call it from a single thread. the scala-native-loop doesn't have any guardrail to avoid that you call the API from different threads.
It might be that the scala native test runner in multithreading mode uses a different thread to run tests, so the event loop is accessed from multiple threads.
If you want to implement a minimal event loop for ZIO you can do with a timer thread and run the sleep callbacks there as Cats effect does, that is better than relying on libuv.
You could vendor some code from cats-effect_

We need to move to a custom event loop to mitigate this flaky behavior of multithreading we observing now. 


> we could also remove installing libuv from the CI jobs.

After this me and @asr2003 have done many dirty tricks that we concluded the culprit was `libuv`. I have isolated the test to n=1(aka single-threaded mode) passes consistently ran 10000 times of this flaky tests and when n>1, no way it reports this flay behavior like once in the moon as we are seeing as the multiple threads are accessing the libuv which isn't thread safe.


And other thing is that we also tried with Single-Threaded Event Loop and designed a simple event loop that runs on a dedicated thread, ensuring that all interactions with libuv are handled by this thread alone and avoid multiple threads accessing libuv. But it also behaves the same. So far I have concluded that we have to move from `libuv`. I think we need to adapt how Cats effect implemented custom event loop for multithreading for native

**Comment by kyri-petrou** [comment]
> So you need to make sure you call it from a single thread. the scala-native-loop doesn't have any guardrail to avoid that you call the API from different threads.

@varshith257 You removed `scala-native-loop`, so it's not the issue

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -227,6 +227,7 @@ jobs:
       - name: Cache scala dependencies
         uses: coursier/cache-action@v6
       - name: Install libuv
+        if: ${{ matrix.platform == 'JS' }}
```
Comment (sideeffffect) [inline .github/workflows/ci.yml]: This is misunderstanding. Please remove "Install libuv" from all steps. We don't need it anymore.

**Review by sideeffffect — CHANGES_REQUESTED** [review]
Please remove "Install libuv" from all steps.

File: .github/workflows/ci.yml (hunk 12 above)
Comment (varshith257) [reply .github/workflows/ci.yml]: Done!

Comment (varshith257) [reply .github/workflows/ci.yml]: But the Interesting here is that before I tested locally yesterday, disabling it and faced test timeout issues. But here not

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -226,8 +226,6 @@ jobs:
           check-latest: true
       - name: Cache scala dependencies
         uses: coursier/cache-action@v6
-      - name: Install libuv
-        run: sudo apt-get update && sudo apt-get install -y libuv1-dev
```
Comment (sideeffffect) [inline .github/workflows/ci.yml:230]: From _all_ steps. This is at many more places.

**Review by sideeffffect — CHANGES_REQUESTED** [review]
Please remove `libuv1-dev` from everywhere.

File: test-sbt/native/src/main/scala/zio/test/sbt/ZTestRunnerNative.scala [hunk 14]
```diff
@@ -138,7 +138,6 @@ sealed class ZTestTask(
         case _                   =>
       }
     }(Unsafe.unsafe)
-    scala.scalanative.loop.EventLoop.run()
```
Comment (kyri-petrou) [inline test-sbt/native/src/main/scala/zio/test/sbt/ZTestRunnerNative.scala:141]: This was the fix to the issue that tests were not being executed and test results were not being reported. You need to change the code so that it now works without the event loop

Comment (varshith257) [reply test-sbt/native/src/main/scala/zio/test/sbt/ZTestRunnerNative.scala:141]: Does it mean we need a custom event loop?

**Comment by varshith257** [comment]
The JS test failure is unrelated to this. I have often faced this test fail. Maybe it's flaky 

https://github.com/zio/zio/actions/runs/10441451696/job/28912596834?pr=9117#step:6:3894

@sideeffffect The issue we encountered till now is due to Ringbuffer being scattered to support the single-threaded operations and we also have expected buffer capacity n>=2. This is the way I haven't seen the `works on large lists` test running when n=1 and the flaky behavior we observed because it needs to create multiple threads and process this large list. and now made thread-safe of ringbuffer

**Comment by varshith257** [comment]
@kyri-petrou I noticed you have also submitted PR for multithreading support for Scala Native and I want to acknowledge the work you've put into it. Over the past four days, I have also been working diligently on this issue, trying different approaches to align with ZIO’s requirements.

I understand that my approach might not have fully aligned with what was needed possibly due to my relative inexperience with certain aspects of ZIO.  It would have been helpful to know earlier that you were also working on this as it might have allowed us to collaborate more effectively from the beginning. 

 Moving forward, I’m open to any opportunity to collaborate with you on this PR.

**Comment by kyri-petrou** [comment]
@varshith257 you opened a PR with 50 LOC and 2 files changed that were not even necessary. And then proceeded to crowd-source through review comments all the changes you did since. Plus, half the tests are not even running in your PR at the moment and they're hanging. I told you multiple times that this PR is far from solving the issue and you should either close it or put it in draft. I'm not sure what you'd like us to collaborate on

**Comment by varshith257** [comment]
> you opened a PR with 50 LOC and 2 files changed that were not even necessary.
Yes, I have already highlighted which was indeed my relative inexperience with all ZIO aspects. And I followed the discussion in Scala 0.5 support and I have seen that copy/pasted scala code and with discussions of enabling and disabling multithreading. So I thought the issue is of to ditch and solve the issues encountered when multithreading is enabled.


You could potentially collaborated with us if you have the good approach what you thought. You have followed up on other reviews till 17 hours ago too which thought we are on track and your last review is of to remove scala-native-loop.

I do argue if you have collaborated on this PR rather going submit another PR against us. Everyone could value their time and efforts right! 

cc: @sideeffffect

**Comment by varshith257** [comment]
> you opened a PR with 50 LOC and 2 files changed that were not even necessary.

Yes, I have already highlighted which was indeed my relative inexperience with all ZIO aspects. And I followed the discussion in Scala 0.5 support and I have seen that copy/pasted scala code and with discussions of enabling and disabling multithreading. So I thought the issue is of to ditch and solve the issues encountered when multithreading is enabled.

> Plus, half the tests are not even running in your PR at the moment and they're hanging.

Check latest changes.


You could potentially collaborated with us if you have the good approach what you thought. You have followed up on other reviews till 17 hours ago too which thought we are on track and your last review is of to remove scala-native-loop.

I do argue if you have collaborated on this PR rather going submit another PR against us. Everyone could value their time and efforts right! 

cc: @sideeffffect

**Comment by kyri-petrou** [comment]
@varshith257 I told you 3 days ago what needed to be done: https://github.com/zio/zio/pull/9117#issuecomment-2293704697, and since then you:

1. Moved only ~1/5th of the code
2. Changed the implementation of `RingBuffer` to something that is still not thread safe and tried to tell us that it is
3. Changed the implementation of `ConcurrentMetricHooksPlatformSpecific` to something that doesn't work because of how Double's are boxed into objects that fail referential equality checks on CAS retries
4. Tried to convince us that the problem with concurrency was due to `libuv` _after_ it was no longer used
5. Tried to convince us multiple times that the tests are working correctly (PS: They're still not!!!)

I was more than happy to give you guidance to work through this issue but I gave up when I realised you had no idea what you were doing. Besides moving a little bit of code, everything you did after 3 days was to either try and convince us that things in this PR are working when they're not or implement things that are not thread-safe/don't work.

**Comment by 987Nabil** [comment]
my 2 cent: I am absolutely happy, if ppl like to learn about zio internals and start contributing. Picking for this one of the most complicated and biggest changes in the zio codebase for a long time, might not be the best idea.
This change is even for experienced zio contributors a hard task.

**Comment by varshith257** [comment]
@987Nabil Yes I agree with you, I have taken and collaborated with @asr2003( who initially attempted this issue) to gain more experience about Zio by contributing to it. It will make more learning experience and also deep dive into Zio internals which can make me get better grasping with fundamental internals of Zio.

Anyhow, I have learnt a lot about Zio by contributing to this issue. It would be happy if @kyri-petrou has collaborated on this PR rather going to submit another PR 😅

**Comment by kyri-petrou** [comment]
@varshith257 I'm sorry to hear that you feel this way. I also apologise for my tone in the previous messages, I've been frustrated with having to repeat myself that the tests are not working and you trying to convince me that they are.

Having said that, you need to understand that I was working on this issue way before you opened a PR, and despite that, I did give you a guys a chance to complete it instead. I only proceeded with opening the PR with the work I did only after it became obvious that this task was way too complex for you. Since I had most of the work done, there was nothing really left to collaborate on. The alternative would have been for me to guide you through reimplementing everything that I had done already, while triple-checking every LOC.

**Comment by jdegoes** [comment]
Closed in favor of @kyri-petrou's PR which was working and more comprehensive.

**Comment by jdegoes** [comment]
@varshith257 Btw, thanks for attempting this. It was really hard! Generally we encourage contributors to pick up something only a little bit beyond their current level, because while we do dedicate time for helping out each contributor, lengthier mentoring or training sessions are very rare. Let me know if you want me to make a ticket specifically for you -- happy to do that!

**Comment by varshith257** [comment]
@jdegoes Thanks for it. I am happier with I have explore bit extent of current experience with zio and learnt a lot about Zio internals working. 

> Let me know if you want me to make a ticket specifically for you -- happy to do that!.

Yes, I am ready to take the challenge and explore more zio while contributing to it
